res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,772 @@
+        off: 0x01,
+        shift: function() {
+            this.group = "[Channel" + number + "]";
+            this.inKey = "pitch_down";
+        },
+        unshift: function() {
+            this.group = "[EffectRack1_EffectUnit" + number + "_Effect1]";
+            this.inKey = "next_effect";
+        }
+    });
+
+    this.beatsnap = new components.Button({
+        type: components.Button.prototype.types.toggle,
+        key: "quantize",
+        midi: [0x89, 0x04],
+        off: 0x01
+    });
+
+    this.reconnectComponents(function(component) {
+        if (component.group === undefined) {
+            component.group = this.currentDeck;
+        }
+    });
+};
+
+MixtrackProFX.Deck.prototype = new components.Deck();
+
+MixtrackProFX.Browse = function() {
+    this.knob = new components.Encoder({
+        group: "[Library]",
+        inKey: "Move",
+        input: function(channel, control, value) {
+            if (value === 0x01) {
+                engine.setParameter(this.group, this.inKey + "Down", 1);
+            } else if (value === 0x7F) {
+                engine.setParameter(this.group, this.inKey + "Up", 1);
+            }
+        }
+    });
+
+    this.knobShift = new components.Encoder({
+        input: function(channel, control, value) {
+            if (value === 0x01) {
+                engine.setParameter("[Channel1]", "waveform_zoom_down", 1);
+
+                // need to zoom both channels if waveform sync is disabled in Mixxx settings.
+                // and when it's enabled then no need to zoom 2nd channel, as it will cause
+                // the zoom to jump 2 levels at once
+                if (!MixtrackProFX.waveformsSynced) {
+                    engine.setParameter("[Channel2]", "waveform_zoom_down", 1);
+                }
+            } else if (value === 0x7F) {
+                engine.setParameter("[Channel1]", "waveform_zoom_up", 1);
+
+                // see above comment
+                if (!MixtrackProFX.waveformsSynced) {
+                    engine.setParameter("[Channel2]", "waveform_zoom_up", 1);
+                }
+            }
+        }
+    });
+
+    this.knobButton = new components.Button({
+        group: "[Library]",
+        inKey: "MoveFocusForward"
+    });
+
+    this.buttonShift = new components.Button({
+        group: "[Library]",
+        inKey: "GoToItem"
+    });
+
+    this.setBeatgrid = new components.Button({
+        group: "[Channel1]",
+        key: "beats_translate_curpos",
+        midi: [0x88, 0x01],
+        off: 0x01,
+        shift: function() {
+            this.group = "[Channel2]";
+        },
+        unshift: function() {
+            this.group = "[Channel1]";
+        }
+    });
+};
+
+MixtrackProFX.Browse.prototype = new components.ComponentContainer();
+
+MixtrackProFX.HeadGain = function() {
+    components.Pot.call(this);
+};
+
+MixtrackProFX.HeadGain.prototype = new components.Pot({
+    group: "[Master]",
+    inKey: "headGain"
+});
+
```

Comment:
Is there a particular reason you created an entirely new Component for that?

Also, if the Master gain knob changes something on the hardware, it should likely not be coupled to the master gain in mixxx, as that will make the knob very difficult to use to adjust the volume (since you are essentially controlling two gain stages at once).

Reply:
I think that was a leftover from the original mapping by @bad1dea5. It worked well so I didn't touch it. I've simplified it now, only xml was needed.

I don't think master gain knob controls something in hardware. When I removed the control for `[Master]` `gain`, turning the master gain knob did not change volume.

Reply:
Now that I look at it, there are more controls that can be mapped directly in xml with no need for script-binding. For example channel volume.

Reply:
mixing XML and JS is confusing to maintain

Reply:
So should I map all controls with script-binding? Even if I'm able to map them just with xml?

Reply:
I'm personally split on this issue. Mapping everything solely in JS might come with some overhead but is easier to maintain in the long term because you know which file to look in if you want to make changes.
In the long term, we want to get rid of the legacy mapping system with XML entirely and just specify the IO parts in JS as well (so you don't have to waste time writing 2000 lines of XML manually). That is why we're pushing to map everything in JS even though it could be mapped in XML. This way, the XML is theoretically obsolete (just not practically because we haven't implemented an API to make signal handlers in pure JS yet).

Reply:
Only crossfader remained mapped directly to xml, rest of the controls is mapped to js now. The reason why I left crossfader is because of the "fader cuts" feature which requires a second inverted input, I don't know how to write that nicely in the component js API).

Reply:
That's okay. I ended up at the same place with my first mapping (Electrix Tweaker) with everything except the crossfader mapped with JS.